File: src/test/java/model/CardTest.java
package model;

import com.aau.wizard.model.CardFactory;
import com.aau.wizard.model.enums.CardSuit;
import com.aau.wizard.model.ICard;
import com.aau.wizard.model.enums.CardType;
import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;

class CardTest {

    @Test
    void testNumberCardCreation() {
        ICard card = CardFactory.createCard(CardSuit.RED, 7);


        assertEquals(CardSuit.RED, card.getSuit());
        assertEquals(7, card.getValue());
        assertEquals(CardType.NUMBER, card.getType());
        assertEquals("RED_7", card.toString());
    }

    @Test
    void testJesterCardCreation() {
        ICard card = CardFactory.createCard(CardSuit.YELLOW, 0);


        assertEquals(CardSuit.YELLOW, card.getSuit());
        assertEquals(0, card.getValue());
        assertEquals(CardType.JESTER, card.getType());
        assertEquals("JESTER", card.toString());
    }

    @Test
    void testWizardCardCreation() {
        ICard card = CardFactory.createCard(CardSuit.BLUE, 14);


        assertEquals(CardSuit.BLUE, card.getSuit());
        assertEquals(14, card.getValue());
        assertEquals(CardType.WIZARD, card.getType());
        assertEquals("WIZARD", card.toString());
    }

    @Test
    void testEdgeCaseNumberCards() {
        ICard lowestNumber = CardFactory.createCard(CardSuit.GREEN, 1);
        ICard highestNumber = CardFactory.createCard(CardSuit.GREEN, 13);

        assertEquals(CardType.NUMBER, lowestNumber.getType());
        assertEquals(CardType.NUMBER, highestNumber.getType());
        assertEquals("GREEN_1", lowestNumber.toString());
        assertEquals("GREEN_13", highestNumber.toString());
    }

    @Test
    void testAllSuitsForNumberCard() {
        for (CardSuit suit : CardSuit.values()) {
            ICard card = CardFactory.createCard(suit, 5);

            assertEquals(suit, card.getSuit());
            assertTrue(card.toString().contains(suit.name()));
        }
    }

    @Test
    void testAllSuitsForJester() {
        for (CardSuit suit : CardSuit.values()) {
            ICard card = CardFactory.createCard(suit, 0);

            assertEquals(suit, card.getSuit());
        }
    }

    @Test
    void testAllSuitsForWizard() {
        for (CardSuit suit : CardSuit.values()) {
            ICard card = CardFactory.createCard(suit, 14);
            assertEquals(suit, card.getSuit());
        }
    }

    @Test
    void testSpecialSuitCards() {
        ICard numberSpecial = CardFactory.createCard(CardSuit.SPECIAL, 5);
        ICard jesterSpecial = CardFactory.createCard(CardSuit.SPECIAL, 0);
        ICard wizardSpecial = CardFactory.createCard(CardSuit.SPECIAL, 14);


        assertEquals("SPECIAL_5", numberSpecial.toString());
        assertEquals("JESTER", jesterSpecial.toString());
        assertEquals("WIZARD", wizardSpecial.toString());
    }

    @Test
    void hashCode_shouldBeEqualForSameCard() {
        ICard card1 = CardFactory.createCard(CardSuit.RED, 5);
        ICard card2 = CardFactory.createCard(CardSuit.RED, 5);

        assertEquals(card1.hashCode(), card2.hashCode(),
                "Two cards with same suit and value should have same hash code");
    }
}

File: src/test/kotlin/com/aau/wizard/core/cards/CardTest.kt
package com.aau.wizard.core.cards

import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.ValueSource
import org.junit.jupiter.api.Nested
import org.junit.jupiter.params.provider.EnumSource

class CardTest {

    @Nested
    inner class CardTypeDetection {
        @Test
        fun `number card has correct type`() {
            val card = Card(Suit.RED, 5)
            assertEquals(CardType.NUMBER, card.type)
        }


        @Test
        fun `jester card has correct type`() {
            val card = Card(Suit.YELLOW, 0)
            assertEquals(CardType.JESTER, card.type)
        }

        @Test
        fun `wizard card has correct type`() {
            val card = Card(Suit.BLUE, 14)
            assertEquals(CardType.WIZARD, card.type)
        }
    }

    @Nested
    inner class CardToString {
        @Test
        fun `number card string representation`() {
            val card = Card(Suit.GREEN, 7)
            assertEquals("7 of green", card.toString())
        }

        @Test
        fun `jester card string representation`() {
            val card = Card(Suit.SPECIAL, 0)
            assertEquals("Jester (special)", card.toString())
        }

        @Test
        fun `wizard card string representation`() {
            val card = Card(Suit.RED, 14)
            assertEquals("Wizard (red)", card.toString())
        }
    }


    @Nested
    inner class CardProperties {
        @ParameterizedTest
        @EnumSource(Suit::class)
        fun `card has correct suit`(suit: Suit) {
            val card = Card(suit, 3)
            assertEquals(suit, card.suit)
        }

        @ParameterizedTest
        @ValueSource(ints = [1, 5, 13])
        fun `number card has correct value`(value: Int) {
            val card = Card(Suit.BLUE, value)
            assertEquals(value, card.value)
            assertEquals(CardType.NUMBER, card.type)
        }

        @Test
        fun `jester has correct value`() {
            val card = Card(Suit.YELLOW, 0)
            assertEquals(0, card.value)
            assertEquals(CardType.JESTER, card.type)
        }

        @Test
        fun `wizard has correct value`() {
            val card = Card(Suit.SPECIAL, 14)
            assertEquals(14, card.value)
            assertEquals(CardType.WIZARD, card.type)
        }
    }

    @Nested
    inner class EdgeCases {
        @Test
        fun `minimum number card value`() {
            val card = Card(Suit.RED, 1)
            assertEquals(1, card.value)
            assertEquals(CardType.NUMBER, card.type)
        }

        @Test
        fun `maximum number card value`() {
            val card = Card(Suit.GREEN, 13)
            assertEquals(13, card.value)
            assertEquals(CardType.NUMBER, card.type)
        }


        @Test
        fun `special suit with number card`() {
            val card = Card(Suit.SPECIAL, 5)
            assertEquals(Suit.SPECIAL, card.suit)
            assertEquals(CardType.NUMBER, card.type)
        }
    }

    @Nested
    inner class ToStringCoverage {
        @Test
        fun `toString for WIZARD card`() {
            val wizardCard = Card(Suit.RED, 14)
            assertEquals("Wizard (red)", wizardCard.toString())
        }

        @Test
        fun `toString for JESTER card`() {
            val jesterCard = Card(Suit.YELLOW, 0)
            assertEquals("Jester (yellow)", jesterCard.toString())
        }

        @Test
        fun `toString for NUMBER card`() {
            val numberCard = Card(Suit.BLUE, 5)
            assertEquals("5 of blue", numberCard.toString())
        }
    }
}

class SuitTest {
    @Test
    fun `all suit values exist`() {
        val values = Suit.values()
        assertEquals(5, values.size)
        assertTrue(values.contains(Suit.RED))
        assertTrue(values.contains(Suit.YELLOW))
        assertTrue(values.contains(Suit.BLUE))
        assertTrue(values.contains(Suit.GREEN))
        assertTrue(values.contains(Suit.SPECIAL))
    }
}


class CardTypeTest {
    @Test
    fun `all card type values exist`() {
        val values = CardType.values()
        assertEquals(3, values.size)
        assertTrue(values.contains(CardType.NUMBER))
        assertTrue(values.contains(CardType.WIZARD))
        assertTrue(values.contains(CardType.JESTER))
    }
}